src/lib/math/pcurves/pcurves_impl/pcurves_impl.h

```diff
@@ -458,6 +458,131 @@ class IntMod final {
       */
       constexpr Self invert() const { return pow_vartime(Self::P_MINUS_2); }
 
+      /**
+      * Helper for variable time BEEA
+      *
+      * Note this function assumes that its arguments are in the standard
+      * domain, not the Montgomery domain. invert_vartime converts its argument
+      * out of Montgomery, and then back to Montgomery when returning the result.
+      */
+      static constexpr void _invert_vartime_div2_helper(Self& a, Self& x) {
+         constexpr auto INV_2 = p_div_2_plus_1(Rep::P);
+
+         for(;;) {
+            if((a.m_val[0] & 1) == 1) {
+               break;
+            }
+
+            shift_right<1>(a.m_val);
+
+            W borrow = shift_right<1>(x.m_val);
+
+            if(borrow) {
+               // Hack to work around GCC miscompilation
```

Comment:
Is there a GCC bug # that should be cited?

Reply:
GCC was miscompiling but given it seemed implicated with inline asm, which is pretty finicky and sometimes just fails when called in a new context even if it has worked fine for years, I was not going to file a bug until I was quite certain it was an actual GCC bug.

It appears the issue is lacking "volatile" on a block of inline asm that (after some nasty macro expansion) is just an 8 word long addition-with-carry chain. Even though the extended asm clobbers memory, GCC seems to not realize that calling it twice might result in different effects.

I had added `volatile` to the other asm statements that write to memory in 9b1b666f81 but apparently missed this one and of course had forgotten about it entirely.